File: lib/iap/native/ios/IAPBridge.swift
//
//  IAPBridge.swift
//  Native half of lib/iap for iOS — StoreKit 2 behind the file-drop protocol
//  the Lua module speaks.
//
//  Integration (see INTEGRATION.md for the full walk-through):
//
//    1. Add this file to the love-ios Xcode project (target: love-ios).
//    2. Add the In-App Purchase capability and the StoreKit framework.
//    3. In AppDelegate.application(_:didFinishLaunchingWithOptions:):
//           IAPBridge.shared.start(loveIdentity: "your_love_identity")
//
//  Requires iOS 15+ (StoreKit 2). Nothing in LÖVE itself needs patching: the
//  two halves only ever meet through files in the game's save directory.
//

import Foundation
import StoreKit

@available(iOS 15.0, *)
public final class IAPBridge {

    public static let shared = IAPBridge()

    private let bridgeDirName = "iap_bridge"
    private let pollInterval: TimeInterval = 0.25

    private var inbox: URL!      // to_native — written by Lua
    private var outbox: URL!     // to_lua    — written by us

    private let queue = DispatchQueue(label: "iap.bridge", qos: .utility)
    private var pollTimer: DispatchSourceTimer?
    private var updatesTask: Task<Void, Never>?

    /// productID -> Product, refreshed on every init.
    private var products: [String: Product] = [:]
    /// Lua's declared type per productID, so finish() knows consume vs. acknowledge.
    private var productTypes: [String: String] = [:]
    /// transaction id (as string) -> Transaction, so a finish request can find it.
    private var liveTransactions: [String: Transaction] = [:]

    private var replySeq: UInt64 = 0
    private var started = false

    private init() {}

    // MARK: - Lifecycle

    public func start(loveIdentity: String) {
        guard !started else { return }
        started = true

        let base = IAPBridge.resolveBridgeDir(loveIdentity: loveIdentity,
                                              dirName: bridgeDirName)
        inbox  = base.appendingPathComponent("to_native")
        outbox = base.appendingPathComponent("to_lua")
        try? FileManager.default.createDirectory(at: inbox,  withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: outbox, withIntermediateDirectories: true)
        NSLog("[IAPBridge] bridge directory: %@", base.path)

        // Transactions can arrive with no purchase in flight: Ask to Buy being
        // approved, a purchase made on another device, a subscription renewing.
        updatesTask = Task.detached(priority: .background) { [weak self] in
            for await result in Transaction.updates {
                await self?.deliver(result, restored: false)
            }
        }

        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now(), repeating: pollInterval)
        timer.setEventHandler { [weak self] in self?.drainInbox() }
        timer.resume()
        pollTimer = timer
    }

    public func stop() {
        pollTimer?.cancel(); pollTimer = nil
        updatesTask?.cancel(); updatesTask = nil
        started = false
    }

    /// LÖVE for iOS keeps save files under `Documents/save/<identity>`. Older
    /// builds used `Library/Application Support/LOVE/<identity>`. Both are
    /// checked; if neither exists yet the Documents location is created and Lua
    /// meets us there. This is the only place the path is decided.
    private static func resolveBridgeDir(loveIdentity: String, dirName: String) -> URL {
        let fm = FileManager.default
        var candidates: [URL] = []

        if let docs = fm.urls(for: .documentDirectory, in: .userDomainMask).first {
            candidates.append(docs.appendingPathComponent("save/\(loveIdentity)/\(dirName)"))
            candidates.append(docs.appendingPathComponent("\(loveIdentity)/\(dirName)"))
        }
        if let support = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
            candidates.append(support.appendingPathComponent("LOVE/\(loveIdentity)/\(dirName)"))
        }

        for url in candidates where fm.fileExists(atPath: url.path) { return url }
        return candidates.first!
    }

    // MARK: - File-drop transport

    private func drainInbox() {
        guard let inbox = inbox else { return }
        let fm = FileManager.default
        guard let names = try? fm.contentsOfDirectory(atPath: inbox.path) else { return }

        // Filenames are ordered by construction, so sorting preserves send order.
        for name in names.sorted() where name.hasSuffix(".rdy") {
            let stem   = String(name.dropLast(4))
            let body   = inbox.appendingPathComponent("\(stem).json")
            let data   = try? Data(contentsOf: body)

            // Consume before handling: a message we cannot parse must not be
            // retried forever.
            try? fm.removeItem(at: inbox.appendingPathComponent(name))
            try? fm.removeItem(at: body)

            guard let data = data,
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                NSLog("[IAPBridge] undecodable request: %@", stem)
                continue
            }
            Task { await handle(obj) }
        }
    }

    /// Write a reply body, then its commit marker. Lua only reads committed pairs.
    private func reply(_ msg: [String: Any]) {
        queue.async {
            self.replySeq += 1
            let stem = String(format: "n-%.0f-%06llu",
                              Date().timeIntervalSince1970 * 1000, self.replySeq)
            let body   = self.outbox.appendingPathComponent("\(stem).json")
            let marker = self.outbox.appendingPathComponent("\(stem).rdy")
            do {
                let data = try JSONSerialization.data(withJSONObject: msg)
                try data.write(to: body, options: .atomic)
                try Data("1".utf8).write(to: marker, options: .atomic)
            } catch {
                NSLog("[IAPBridge] reply failed: %@", String(describing: error))
                try? FileManager.default.removeItem(at: body)
            }
        }
    }

    private func replyError(id: String?, op: String, code: String, message: String) {
        var msg: [String: Any] = ["op": op, "ok": false, "code": code, "message": message]
        if let id = id { msg["id"] = id }
        reply(msg)
    }

    // MARK: - Request handling

    private func handle(_ msg: [String: Any]) async {
        let op = msg["op"] as? String ?? ""
        let id = msg["id"] as? String

        switch op {
        case "init":     await doInit(id: id, products: msg["products"] as? [[String: Any]] ?? [])
        case "purchase": await doPurchase(id: id, sku: msg["sku"] as? String ?? "")
        case "restore":  await doRestore(id: id)
        case "finish":   await doFinish(id: id, msg: msg)
        default:         NSLog("[IAPBridge] unknown op: %@", op)
        }
    }

    private func doInit(id: String?, products declared: [[String: Any]]) async {
        productTypes.removeAll()
        var ids: [String] = []
        for entry in declared {
            guard let sku = entry["sku"] as? String, !sku.isEmpty else { continue }
            ids.append(sku)
            productTypes[sku] = entry["type"] as? String ?? "consumable"
        }

        var details: [[String: Any]] = []
        do {
            let fetched = try await Product.products(for: ids)
            self.products = Dictionary(uniqueKeysWithValues: fetched.map { ($0.id, $0) })
            details = fetched.map { product in
                [
                    "sku": product.id,
                    "price": product.displayPrice,
                    "priceAmountMicros": NSDecimalNumber(decimal: product.price * 1_000_000).int64Value,
                    "currency": product.priceFormatStyle.currencyCode,
                    "title": product.displayName,
                    "description": product.description,
                ]
            }
        } catch {
            replyError(id: id, op: "init", code: "service_unavailable",
                       message: String(describing: error))
            return
        }

        // Anything StoreKit still considers unfinished is money we may owe the
        // player: hand it to Lua, which grants it and then asks us to finish.
        var unfinished: [[String: Any]] = []
        for await result in Transaction.unfinished {
            if let entry = record(result, restored: false) { unfinished.append(entry) }
        }

        var msg: [String: Any] = ["op": "init", "ok": true,
                                  "products": details, "unfinished": unfinished]
        if let id = id { msg["id"] = id }
        reply(msg)
    }

    private func doPurchase(id: String?, sku: String) async {
        guard let product = products[sku] else {
            replyError(id: id, op: "purchase", code: "product_unavailable",
                       message: "No product details for \(sku)")
            return
        }

        do {
            switch try await product.purchase() {
            case .success(let verification):
                guard let entry = record(verification, restored: false) else {
                    // StoreKit could not verify its own signature — do not
                    // finish it, and never grant it.
                    replyError(id: id, op: "purchase", code: "invalid_receipt",
                               message: "Transaction failed StoreKit verification")
                    return
                }
                reply(["op": "purchase", "ok": true, "purchase": entry])

            case .userCancelled:
                replyError(id: id, op: "purchase", code: "user_cancelled",
                           message: "Cancelled by the player")

            case .pending:
                // Ask to Buy / SCA: nothing yet, it arrives via Transaction.updates.
                reply(["op": "purchase_deferred", "sku": sku])

            @unknown default:
                replyError(id: id, op: "purchase", code: "store_error",
                           message: "Unknown StoreKit result")
            }
        } catch {
            replyError(id: id, op: "purchase", code: "store_error",
                       message: String(describing: error))
        }
    }

    private func doRestore(id: String?) async {
        // Pull anything the account is entitled to. syncing first makes the
        // "Restore Purchases" button behave the way reviewers expect.
        try? await AppStore.sync()

        var list: [[String: Any]] = []
        for await result in Transaction.currentEntitlements {
            if let entry = record(result, restored: true) { list.append(entry) }
        }

        var msg: [String: Any] = ["op": "restore", "ok": true, "purchases": list]
        if let id = id { msg["id"] = id }
        reply(msg)
    }

    /// Lua has granted the purchase; tell StoreKit we are done with it.
    /// StoreKit 2 makes no consume/acknowledge distinction — finish() covers both.
    private func doFinish(id: String?, msg: [String: Any]) async {
        let token = msg["token"] as? String ?? ""
        let txn   = msg["txn"] as? String

        var finished = false
        if let transaction = liveTransactions[token] {
            await transaction.finish()
            liveTransactions.removeValue(forKey: token)
            finished = true
        } else {
            // Not in memory (a relaunch, say): find it among the unfinished set.
            for await result in Transaction.unfinished {
                if case .verified(let transaction) = result,
                   String(transaction.id) == token {
                    await transaction.finish()
                    finished = true
                    break
                }
            }
            // Nothing left to finish is success as far as Lua is concerned.
            if !finished { finished = true }
        }

        var out: [String: Any] = ["op": "finish", "ok": finished]
        if let id = id   { out["id"] = id }
        if let txn = txn { out["txn"] = txn }
        reply(out)
    }

    // MARK: - Transactions

    private func deliver(_ result: VerificationResult<Transaction>, restored: Bool) async {
        guard let entry = record(result, restored: restored) else {
            NSLog("[IAPBridge] dropped an unverified transaction")
            return
        }
        reply(["op": "purchase", "ok": true, "purchase": entry])
    }

    /// Turn a verified StoreKit transaction into the Lua-side purchase shape.
    /// Unverified results return nil: they are never granted and never finished.
    private func record(_ result: VerificationResult<Transaction>,
                        restored: Bool) -> [String: Any]? {
        guard case .verified(let transaction) = result else { return nil }

        let token = String(transaction.id)
        liveTransactions[token] = transaction

        return [
            "sku": transaction.productID,
            "txn": token,
            "token": token,
            // The signed JWS. Send this to your own server and verify it there
            // with the App Store Server API — never trust the device.
            "payload": result.jwsRepresentation,
            "signature": "",
            "platform": "ios",
            "restored": restored,
        ]
    }
}

File: lib/iap/native/android/IAPBridge.java
package org.love2d.android;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import com.android.billingclient.api.AcknowledgePurchaseParams;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ConsumeParams;
import com.android.billingclient.api.PendingPurchasesParams;
import com.android.billingclient.api.ProductDetails;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.QueryProductDetailsParams;
import com.android.billingclient.api.QueryPurchasesParams;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Native half of lib/iap for Android — Google Play Billing behind the file-drop
 * protocol the Lua module speaks.
 *
 * Integration (see INTEGRATION.md for the full walk-through):
 *
 *   app/build.gradle:
 *     implementation 'com.android.billingclient:billing:7.1.1'
 *
 *   GameActivity.onCreate():
 *     IAPBridge.start(this, "your_love_identity");   // t.identity from conf.lua
 *   GameActivity.onDestroy():
 *     IAPBridge.stop();
 *
 * No changes to LÖVE itself are required: the two sides only ever meet through
 * files in the game's save directory.
 */
public final class IAPBridge implements PurchasesUpdatedListener {

    private static final String TAG = "IAPBridge";
    private static final String BRIDGE_DIR = "iap_bridge";
    private static final long   POLL_MS = 250L;
    private static final Charset UTF8 = Charset.forName("UTF-8");

    private static IAPBridge instance;

    private final Activity activity;
    private final File inbox;      // to_native — written by Lua
    private final File outbox;     // to_lua    — written by us

    private final HandlerThread pollThread;
    private final Handler pollHandler;
    private final Handler mainHandler;

    private BillingClient billing;
    private final AtomicBoolean connected = new AtomicBoolean(false);
    private final AtomicBoolean stopped   = new AtomicBoolean(false);

    /** sku -> details, refreshed on every init. */
    private final Map<String, ProductDetails> details = new HashMap<>();
    /** sku -> "consumable" | "non_consumable" | "subscription", as Lua declared it. */
    private final Map<String, String> productTypes = new HashMap<>();
    /** purchaseToken -> purchase, so a finish request can find its purchase. */
    private final Map<String, Purchase> knownPurchases = new HashMap<>();

    private long replySeq = 0;

    // ── lifecycle ────────────────────────────────────────────────────────────

    public static synchronized void start(Activity activity, String loveIdentity) {
        if (instance != null) return;
        instance = new IAPBridge(activity, loveIdentity);
        instance.begin();
    }

    public static synchronized void stop() {
        if (instance == null) return;
        instance.end();
        instance = null;
    }

    private IAPBridge(Activity activity, String loveIdentity) {
        this.activity = activity;

        File base = resolveBridgeDir(activity, loveIdentity);
        this.inbox  = new File(base, "to_native");
        this.outbox = new File(base, "to_lua");
        //noinspection ResultOfMethodCallIgnored
        this.inbox.mkdirs();
        //noinspection ResultOfMethodCallIgnored
        this.outbox.mkdirs();

        Log.i(TAG, "bridge directory: " + base.getAbsolutePath());

        this.pollThread = new HandlerThread("iap-bridge-poll");
        this.pollThread.start();
        this.pollHandler = new Handler(pollThread.getLooper());
        this.mainHandler = new Handler(Looper.getMainLooper());
    }

    /**
     * LÖVE for Android stores save files under {@code getFilesDir()/save/<identity>}
     * when {@code t.externalstorage} is false (the default). If your build differs,
     * pass the identity you use, or edit this method — it is the only place the
     * path is decided.
     */
    private static File resolveBridgeDir(Context context, String loveIdentity) {
        File internal = new File(new File(context.getFilesDir(), "save"), loveIdentity);
        File candidate = new File(internal, BRIDGE_DIR);
        if (candidate.isDirectory()) return candidate;

        // t.externalstorage = true
        File external = context.getExternalFilesDir(null);
        if (external != null) {
            File ext = new File(new File(new File(external, "save"), loveIdentity), BRIDGE_DIR);
            if (ext.isDirectory()) return ext;
        }

        // Lua has not run yet; create the internal-storage location and let it
        // meet us there.
        return candidate;
    }

    private void begin() {
        billing = BillingClient.newBuilder(activity)
                .setListener(this)
                .enablePendingPurchases(
                        PendingPurchasesParams.newBuilder().enableOneTimeProducts().build())
                .build();
        connect();
        pollHandler.post(pollLoop);
    }

    private void end() {
        stopped.set(true);
        pollHandler.removeCallbacksAndMessages(null);
        pollThread.quitSafely();
        if (billing != null) {
            billing.endConnection();
            billing = null;
        }
    }

    private void connect() {
        billing.startConnection(new BillingClientStateListener() {
            @Override public void onBillingSetupFinished(@NonNull BillingResult result) {
                connected.set(result.getResponseCode() == BillingClient.BillingResponseCode.OK);
                Log.i(TAG, "billing setup: " + result.getResponseCode() + " " + result.getDebugMessage());
            }
            @Override public void onBillingServiceDisconnected() {
                connected.set(false);
                Log.w(TAG, "billing service disconnected");
                // Lua retries init on its own schedule; reconnect so the retry lands.
                mainHandler.postDelayed(() -> { if (!stopped.get()) connect(); }, 2000L);
            }
        });
    }

    // ── file-drop transport ──────────────────────────────────────────────────

    private final Runnable pollLoop = new Runnable() {
        @Override public void run() {
            if (stopped.get()) return;
            try {
                drainInbox();
            } catch (Throwable t) {
                Log.e(TAG, "poll failed", t);
            }
            pollHandler.postDelayed(this, POLL_MS);
        }
    };

    private void drainInbox() {
        String[] names = inbox.list();
        if (names == null || names.length == 0) return;

        Arrays.sort(names);   // filenames are ordered by construction
        for (String name : names) {
            if (!name.endsWith(".rdy")) continue;

            String stem = name.substring(0, name.length() - 4);
            File body   = new File(inbox, stem + ".json");
            String json = readAll(body);

            // Consume before handling: a message we cannot parse must not be
            // retried forever.
            //noinspection ResultOfMethodCallIgnored
            new File(inbox, name).delete();
            //noinspection ResultOfMethodCallIgnored
            body.delete();

            if (json == null) continue;
            try {
                final JSONObject msg = new JSONObject(json);
                mainHandler.post(() -> handle(msg));   // billing wants the main thread
            } catch (JSONException e) {
                Log.e(TAG, "undecodable request: " + json, e);
            }
        }
    }

    private static String readAll(File f) {
        if (!f.isFile()) return null;
        RandomAccessFile raf = null;
        try {
            raf = new RandomAccessFile(f, "r");
            byte[] buf = new byte[(int) raf.length()];
            raf.readFully(buf);
            return new String(buf, UTF8);
        } catch (IOException e) {
            Log.e(TAG, "read failed: " + f, e);
            return null;
        } finally {
            if (raf != null) try { raf.close(); } catch (IOException ignored) { }
        }
    }

    /** Write a reply body, then its commit marker. Lua only reads committed pairs. */
    private synchronized void reply(JSONObject msg) {
        String stem = String.format("n-%d-%06d", System.currentTimeMillis(), ++replySeq);
        File body   = new File(outbox, stem + ".json");
        File marker = new File(outbox, stem + ".rdy");
        try {
            FileOutputStream out = new FileOutputStream(body);
            out.write(msg.toString().getBytes(UTF8));
            out.flush();
            out.getFD().sync();          // durable before the marker appears
            out.close();

            FileOutputStream mk = new FileOutputStream(marker);
            mk.write('1');
            mk.flush();
            mk.getFD().sync();
            mk.close();
        } catch (IOException e) {
            Log.e(TAG, "reply failed", e);
            //noinspection ResultOfMethodCallIgnored
            body.delete();
        }
    }

    private void replyError(String id, String op, String code, String message) {
        try {
            JSONObject o = new JSONObject();
            if (id != null) o.put("id", id);
            o.put("op", op).put("ok", false).put("code", code).put("message", message);
            reply(o);
        } catch (JSONException e) {
            Log.e(TAG, "replyError failed", e);
        }
    }

    // ── request handling ─────────────────────────────────────────────────────

    private void handle(JSONObject msg) {
        String op = msg.optString("op");
        String id = msg.has("id") ? msg.optString("id") : null;

        if ("init".equals(op))          doInit(id, msg.optJSONArray("products"));
        else if ("purchase".equals(op)) doPurchase(id, msg.optString("sku"));
        else if ("restore".equals(op))  doRestore(id);
        else if ("finish".equals(op))   doFinish(id, msg);
        else Log.w(TAG, "unknown op: " + op);
    }

    private void doInit(final String id, JSONArray products) {
        if (!connected.get()) {
            replyError(id, "init", "billing_unavailable", "Play Billing is not connected");
            connect();
            return;
        }

        productTypes.clear();
        final List<QueryProductDetailsParams.Product> inapp = new ArrayList<>();
        final List<QueryProductDetailsParams.Product> subs  = new ArrayList<>();

        for (int i = 0; products != null && i < products.length(); i++) {
            JSONObject p = products.optJSONObject(i);
            if (p == null) continue;
            String sku  = p.optString("sku");
            String type = p.optString("type", "consumable");
            if (sku.isEmpty()) continue;
            productTypes.put(sku, type);

            boolean isSub = "subscription".equals(type);
            QueryProductDetailsParams.Product entry = QueryProductDetailsParams.Product.newBuilder()
                    .setProductId(sku)
                    .setProductType(isSub ? BillingClient.ProductType.SUBS : BillingClient.ProductType.INAPP)
                    .build();
            (isSub ? subs : inapp).add(entry);
        }

        final JSONArray found = new JSONArray();
        final int[] remaining = { (inapp.isEmpty() ? 0 : 1) + (subs.isEmpty() ? 0 : 1) };
        if (remaining[0] == 0) {
            finishInit(id, found);
            return;
        }

        details.clear();
        for (final List<QueryProductDetailsParams.Product> batch : Arrays.asList(inapp, subs)) {
            if (batch.isEmpty()) continue;
            QueryProductDetailsParams params =
                    QueryProductDetailsParams.newBuilder().setProductList(batch).build();

            billing.queryProductDetailsAsync(params, (result, list) -> {
                if (result.getResponseCode() == BillingClient.BillingResponseCode.OK && list != null) {
                    for (ProductDetails pd : list) {
                        details.put(pd.getProductId(), pd);
                        JSONObject entry = describe(pd);
                        if (entry != null) found.put(entry);
                    }
                } else {
                    Log.w(TAG, "queryProductDetails: " + result.getResponseCode()
                            + " " + result.getDebugMessage());
                }
                if (--remaining[0] == 0) finishInit(id, found);
            });
        }
    }

    private static JSONObject describe(ProductDetails pd) {
        try {
            JSONObject o = new JSONObject();
            o.put("sku", pd.getProductId());
            o.put("title", pd.getTitle());
            o.put("description", pd.getDescription());

            ProductDetails.OneTimePurchaseOfferDetails one = pd.getOneTimePurchaseOfferDetails();
            if (one != null) {
                o.put("price", one.getFormattedPrice());
                o.put("priceAmountMicros", one.getPriceAmountMicros());
                o.put("currency", one.getPriceCurrencyCode());
                return o;
            }

            List<ProductDetails.SubscriptionOfferDetails> offers = pd.getSubscriptionOfferDetails();
            if (offers != null && !offers.isEmpty()) {
                List<ProductDetails.PricingPhase> phases =
                        offers.get(0).getPricingPhases().getPricingPhaseList();
                if (!phases.isEmpty()) {
                    ProductDetails.PricingPhase phase = phases.get(phases.size() - 1);
                    o.put("price", phase.getFormattedPrice());
                    o.put("priceAmountMicros", phase.getPriceAmountMicros());
                    o.put("currency", phase.getPriceCurrencyCode());
                }
            }
            return o;
        } catch (JSONException e) {
            Log.e(TAG, "describe failed", e);
            return null;
        }
    }

    /** Report product details plus everything Play still considers unfinished. */
    private void finishInit(final String id, final JSONArray products) {
        queryOwned((purchases, ok) -> {
            try {
                JSONArray unfinished = new JSONArray();
                for (Purchase p : purchases) {
                    // PENDING purchases carry no entitlement yet; Play delivers
                    // them again once they clear.
                    if (p.getPurchaseState() != Purchase.PurchaseState.PURCHASED) continue;
                    knownPurchases.put(p.getPurchaseToken(), p);
                    for (JSONObject entry : toJson(p, false)) unfinished.put(entry);
                }
                JSONObject o = new JSONObject();
                if (id != null) o.put("id", id);
                o.put("op", "init").put("ok", true)
                 .put("products", products).put("unfinished", unfinished);
                reply(o);
            } catch (JSONException e) {
                replyError(id, "init", "internal", e.getMessage());
            }
        });
    }

    private void doPurchase(final String id, final String sku) {
        ProductDetails pd = details.get(sku);
        if (pd == null) {
            replyError(id, "purchase", "product_unavailable", "No details for " + sku);
            return;
        }

        BillingFlowParams.ProductDetailsParams.Builder pdp =
                BillingFlowParams.ProductDetailsParams.newBuilder().setProductDetails(pd);

        List<ProductDetails.SubscriptionOfferDetails> offers = pd.getSubscriptionOfferDetails();
        if (offers != null && !offers.isEmpty()) {
            pdp.setOfferToken(offers.get(0).getOfferToken());
        }

        BillingFlowParams flow = BillingFlowParams.newBuilder()
                .setProductDetailsParamsList(java.util.Collections.singletonList(pdp.build()))
                .build();

        BillingResult result = billing.launchBillingFlow(activity, flow);
        if (result.getResponseCode() != BillingClient.BillingResponseCode.OK) {
            replyError(id, "purchase", codeFor(result.getResponseCode()), result.getDebugMessage());
        }
        // Success is delivered asynchronously through onPurchasesUpdated.
    }

    private void doRestore(final String id) {
        queryOwned((purchases, ok) -> {
            try {
                JSONArray list = new JSONArray();
                for (Purchase p : purchases) {
                    if (p.getPurchaseState() != Purchase.PurchaseState.PURCHASED) continue;
                    knownPurchases.put(p.getPurchaseToken(), p);
                    for (JSONObject entry : toJson(p, true)) list.put(entry);
                }
                JSONObject o = new JSONObject();
                if (id != null) o.put("id", id);
                o.put("op", "restore").put("ok", ok).put("purchases", list);
                reply(o);
            } catch (JSONException e) {
                replyError(id, "restore", "internal", e.getMessage());
            }
        });
    }

    /**
     * Settle a purchase Lua has already granted: consumables are consumed so
     * they can be bought again, everything else is acknowledged (Play refunds
     * anything left unacknowledged for three days).
     */
    private void doFinish(final String id, JSONObject msg) {
        final String token = msg.optString("token");
        final String txn   = msg.optString("txn");
        final String type  = msg.optString("type", "consumable");

        if (token.isEmpty()) {
            replyError(id, "finish", "no_token", "finish request carried no purchase token");
            return;
        }

        if ("consumable".equals(type)) {
            ConsumeParams params = ConsumeParams.newBuilder().setPurchaseToken(token).build();
            billing.consumeAsync(params, (result, outToken) ->
                    replyFinish(id, txn, result.getResponseCode() == BillingClient.BillingResponseCode.OK
                                    || result.getResponseCode() == BillingClient.BillingResponseCode.ITEM_NOT_OWNED,
                            result));
            return;
        }

        Purchase known = knownPurchases.get(token);
        if (known != null && known.isAcknowledged()) {
            replyFinish(id, txn, true, null);
            return;
        }

        AcknowledgePurchaseParams params = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(token).build();
        billing.acknowledgePurchase(params, result ->
                replyFinish(id, txn, result.getResponseCode() == BillingClient.BillingResponseCode.OK,
                        result));
    }

    private void replyFinish(String id, String txn, boolean ok, BillingResult result) {
        try {
            JSONObject o = new JSONObject();
            if (id != null)  o.put("id", id);
            if (txn != null && !txn.isEmpty()) o.put("txn", txn);
            o.put("op", "finish").put("ok", ok);
            if (!ok && result != null) {
                o.put("code", codeFor(result.getResponseCode()));
                o.put("message", result.getDebugMessage());
            }
            reply(o);
        } catch (JSONException e) {
            Log.e(TAG, "replyFinish failed", e);
        }
    }

    // ── Play callbacks ───────────────────────────────────────────────────────

    @Override
    public void onPurchasesUpdated(@NonNull BillingResult result, List<Purchase> purchases) {
        int code = result.getResponseCode();

        if (code == BillingClient.BillingResponseCode.USER_CANCELED) {
            replyError(null, "purchase", "user_cancelled", "Cancelled by the player");
            return;
        }
        if (code != BillingClient.BillingResponseCode.OK || purchases == null) {
            replyError(null, "purchase", codeFor(code), result.getDebugMessage());
            return;
        }

        for (Purchase p : purchases) {
            if (p.getPurchaseState() == Purchase.PurchaseState.PENDING) {
                try {
                    JSONObject o = new JSONObject();
                    o.put("op", "purchase_deferred");
                    if (!p.getProducts().isEmpty()) o.put("sku", p.getProducts().get(0));
                    reply(o);
                } catch (JSONException ignored) { }
                continue;
            }
            if (p.getPurchaseState() != Purchase.PurchaseState.PURCHASED) continue;

            knownPurchases.put(p.getPurchaseToken(), p);
            for (JSONObject entry : toJson(p, false)) {
                try {
                    reply(new JSONObject().put("op", "purchase").put("ok", true).put("purchase", entry));
                } catch (JSONException e) {
                    Log.e(TAG, "purchase reply failed", e);
                }
            }
        }
    }

    private interface OwnedCallback { void onOwned(List<Purchase> purchases, boolean ok); }

    /** Play keeps INAPP and SUBS in separate ledgers; ask both and merge. */
    private void queryOwned(final OwnedCallback cb) {
        if (!connected.get()) {
            cb.onOwned(new ArrayList<>(), false);
            return;
        }

        final List<Purchase> all = new ArrayList<>();
        final boolean[] ok = { true };
        final int[] remaining = { 2 };

        for (final String type : new String[] { BillingClient.ProductType.INAPP,
                                                BillingClient.ProductType.SUBS }) {
            billing.queryPurchasesAsync(
                    QueryPurchasesParams.newBuilder().setProductType(type).build(),
                    (result, list) -> {
                        if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                            if (list != null) all.addAll(list);
                        } else {
                            ok[0] = false;
                            Log.w(TAG, "queryPurchases(" + type + "): " + result.getDebugMessage());
                        }
                        if (--remaining[0] == 0) cb.onOwned(all, ok[0]);
                    });
        }
    }

    /**
     * A Play purchase can cover several products (multi-line checkout), so this
     * yields one Lua-side purchase per product. The transaction id stays unique
     * per line by suffixing the sku when there is more than one.
     */
    private List<JSONObject> toJson(Purchase p, boolean restored) {
        List<JSONObject> out = new ArrayList<>();
        List<String> skus = p.getProducts();
        String orderId = p.getOrderId();
        if (orderId == null || orderId.isEmpty()) orderId = p.getPurchaseToken();

        for (String sku : skus) {
            try {
                JSONObject o = new JSONObject();
                o.put("sku", sku);
                o.put("txn", skus.size() > 1 ? orderId + ":" + sku : orderId);
                o.put("token", p.getPurchaseToken());
                o.put("payload", p.getOriginalJson());     // verify this server-side
                o.put("signature", p.getSignature());
                o.put("platform", "android");
                o.put("restored", restored);
                out.add(o);
            } catch (JSONException e) {
                Log.e(TAG, "toJson failed", e);
            }
        }
        return out;
    }

    private static String codeFor(int responseCode) {
        switch (responseCode) {
            case BillingClient.BillingResponseCode.USER_CANCELED:        return "user_cancelled";
            case BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED:   return "already_owned";
            case BillingClient.BillingResponseCode.ITEM_UNAVAILABLE:     return "product_unavailable";
            case BillingClient.BillingResponseCode.SERVICE_DISCONNECTED:
            case BillingClient.BillingResponseCode.SERVICE_UNAVAILABLE:  return "service_unavailable";
            case BillingClient.BillingResponseCode.BILLING_UNAVAILABLE:  return "billing_unavailable";
            case BillingClient.BillingResponseCode.NETWORK_ERROR:        return "network_error";
            case BillingClient.BillingResponseCode.DEVELOPER_ERROR:      return "developer_error";
            default:                                                     return "store_error";
        }
    }
}
